button-merchant/src/main/java/com/usebutton/merchant/PostOrderTask.java

```diff
@@ -37,22 +41,75 @@
     private final String sourceToken;
     private final Order order;
     private final DeviceManager deviceManager;
+    private final ThreadManager threadManager;
+
+    private Getter<Double> retryDelayInterval = new Getter<Double>() {
+        @Override
+        public Double get() {
+            return Math.pow(2, retryCount) * 100;
+        }
+    };
+
+    static final int MAX_RETRIES = 4;
+
+    private int retryCount = 0;
 
     PostOrderTask(@Nullable Listener listener, ButtonApi buttonApi, Order order,
-            String applicationId, String sourceToken, DeviceManager deviceManager) {
+            String applicationId, String sourceToken, DeviceManager deviceManager,
+            ThreadManager threadManager) {
         super(listener);
         this.buttonApi = buttonApi;
         this.order = order;
         this.applicationId = applicationId;
         this.sourceToken = sourceToken;
         this.deviceManager = deviceManager;
+        this.threadManager = threadManager;
     }
 
     @Nullable
     @Override
     Void execute() throws Exception {
         String advertisingId = deviceManager.isLimitAdTrackingEnabled()
                 ? null : deviceManager.getAdvertisingId();
-        return buttonApi.postOrder(order, applicationId, sourceToken, advertisingId);
+
+        // loop and execute postOrder until max retries is met or non case exception is met
+        while (true) {
+            try {
+                return buttonApi.postOrder(order, applicationId, sourceToken, advertisingId);
+            } catch (ButtonNetworkException exception) {
+                if (!shouldRetry(exception)) {
+                    throw exception;
+                }
+
+                retryCount++;
+            }
+        }
+    }
+
+    /**
+     * @param exception exception thrown by api request
+     * @return true if should retry
+     */
+    private boolean shouldRetry(ButtonNetworkException exception) throws InterruptedException {
+        if (retryCount >= MAX_RETRIES) {
+            return false;
+        }
+
+        double delay = retryDelayInterval.get();
+        if (exception instanceof HttpStatusException) {
+            HttpStatusException httpStatusException =
+                    (HttpStatusException) exception;
+            if (httpStatusException.wasRateLimited() || httpStatusException.wasServerError()) {
+                threadManager.sleep((long) delay);
+                return true;
```

Comment:
I would think we specifically do _not_ want to retry if the request was rate limited. The purpose of a rate limit is to reduce load on the server. Retrying a rate limited request would have the opposite effect.

Reply:
> Some minor changes requested, otherwise LGTM. In the future if we were to add another top-level API method (like `postOrder()`) we should look into refactoring the retry metadata into its own class (e.g. `RetryConfig`) which a `RetryHandler` can then use. This way that retry logic is abstracted away from the `PostOrderTask`.

Agreed